core/consensus/babe/src/digest.rs

```diff
@@ -0,0 +1,138 @@
+use primitives::sr25519::{Public, Signature};
+use babe_primitives::BABE_ENGINE_ID;
+use runtime_primitives::generic::DigestItem;
+use std::fmt::Debug;
+use parity_codec::{Decode, Encode, Input};
+use log::{debug, info, error};
+use schnorrkel::{
+	vrf::{VRFProof, VRFOutput, VRF_OUTPUT_LENGTH, VRF_PROOF_LENGTH},
+	PUBLIC_KEY_LENGTH,
+};
+
+/// A BABE pre-digest.  It includes:
+///
+/// * The public key of the author.
+/// * The VRF proof.
+/// * The VRF output.
+/// * The slot number.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BabePreDigest {
+	pub(super) vrf_output: VRFOutput,
+	pub(super) proof: VRFProof,
+	pub(super) author: Public,
+	pub(super) slot_num: u64,
+}
+
+/// The prefix used by BABE for its VRF keys.
+pub const BABE_VRF_PREFIX: &'static [u8] = b"substrate-babe-vrf";
+
+macro_rules! babe_assert_eq {
+	($a: expr, $b: expr) => {
+		{
+			let ref a = $a;
+			let ref b = $b;
+			if a != b {
+				error!(
+					target: "babe",
+					"Expected {:?} to equal {:?}, but they were not",
+					stringify!($a),
+					stringify!($b),
+				);
+				assert_eq!(a, b);
+			}
+		}
+	};
+}
+
+type TmpDecode = (
+	[u8; VRF_OUTPUT_LENGTH],
+	[u8; VRF_PROOF_LENGTH],
+	[u8; PUBLIC_KEY_LENGTH],
+	u64,
+);
+
+impl Encode for BabePreDigest {
+	fn encode(&self) -> Vec<u8> {
+		let tmp: TmpDecode = (
+			*self.vrf_output.as_bytes(),
+			self.proof.to_bytes(),
+			self.author.0,
+			self.slot_num,
+		);
+		let encoded = parity_codec::Encode::encode(&tmp);
+		if cfg!(any(test, debug_assertions)) {
```

Comment:
I'm concerned that it has a probability of having zero coverage actaully. If there is no test code in current hitting those lines it will never run. Note that `#[cfg(test)]` is only run when you run test for the crate not if it's only a dependency of other crates.
`debug_assertions` if I understand correctly will be triggered when we run non-release build which is also very questionable: no one will run no-debug substrate code and I believe we used to run tests in `--release` as well.

So IMHO there is a good chance it's introducing a dead code that just affects readability of the implementation.

Reply:
This should be a test.

Reply:
This code will be hit by the BABE integration tests, so essentially the entire BABE test suite is a test case for it.  What is the best way to generate test cases?